### DISCUSSION TITLE
Format decimal field on Datagrid in Edit Form

### DISCUSSION THREAD

**Discussion by bkea** [Megabit/Blazorise#4538] [Q&A] [answered]
I have columns on my Datagrid object of type decimal.  I can enter values like 23.5 or 10.1111 but I cannot enter a value that has a zero on the end or in the middle in the decimal positions (7.10, 7.101, 7.1101).  Each value stops at the number before the zero
7.10 sops at 7.1
7.101 stops at 7.1 

I can enter 7.111 and then change the middle number to zero to fake out the system 7.111 becomes 7.101.  I believe it is a formatting issue, but I have tried a few different ways to get around it unsuccessfully.

```
<DataGridNumericColumn Field="@nameof(CaliberInfo.oal_max)" Caption="OAL Max" DisplayFormat="{0:00.000}"
DisplayFormatProvider="@System.Globalization.CultureInfo.GetCultureInfo("en-EN")" Editable>
<EditTemplate>
<NumericEdit TValue="decimal" Value="@((decimal)(((CellEditContext)context).CellValue))" ValueChanged="@(v=>((CellEditContext)context).CellValue=v)" />
</EditTemplate>
</DataGridNumericColumn>
<DataGridNumericColumn Field="@nameof(CaliberInfo.oal_min)" Caption="OAL Min" DisplayFormat="{0:00.000}"
DisplayFormatProvider="@System.Globalization.CultureInfo.GetCultureInfo("en-EN")" Editable>
<EditTemplate>
<NumericEdit TValue="decimal" Value="@((decimal)(((CellEditContext)context).CellValue))" ValueChanged="@(v=>((CellEditContext)context).CellValue=v)" />
</EditTemplate>
```

Here are two fields both defined as decimal with a format of {0:00.000}. 

I have tried:
making them a DataGridColumn with the same type and format and without an Edit Template
making them a DataGridColumn with the same type and format and with an Edit Template
making them a DataGridNumbericColumn without an Edit Template
making them a DataGridNumbericColumn with an Edit Template (as seen above)
making the format{0:00.###}            //I would think this would not allow zeros on the end.

The data model has these fields defined as decimal.  Everything else works; I just can't enter values like 3.205.

Thank you in advance.

**Comment by stsrki** [comment]
The NumericEdit is a basic input element that doesn't have many options to format the value. Instead, you can use the NumericPicker. It has additional parameters like `RoundingMethod`, or `AllowDecimalPadding` to control the number formating.

**Reply by bkea** [reply]
Thank you @stsrki .  I actually like the number picker over the NumericEdit because the edit invokes the LoadingIndicator after each value I enter.  

Unfortunately, I am still not able to enter a value out to the thousandths position (00.000).  I am able to enter a value out to the hundredths position (99.99).  At first, I kept the definition simple:

<NumericPicker TValue="decimal?" />

When that didn't work, I tried several alternate iterations and definitions including using float instead of decimal and using the Decimal parameter:
<NumericPicker TValue="decimal?" Decimals="4" />

I wanted to try as many combinations as possible before replying.  Unfortunately, I still can't get a value I need.

```
<DataGridNumericColumn Field="@nameof(CaliberInfo.oal_max)" Caption="OAL Max" DisplayFormat="{0:#0.00#}" DisplayFormatProvider="@System.Globalization.CultureInfo.GetCultureInfo("en-EN")" Editable>
    <NumericPicker TValue="decimal?" Decimals="4" AllowDecimalPadding="NumericAllowDecimalPadding.Always"/>
</DataGridNumericColumn>
```

**Comment by bkea** [comment, marked as answer]
FIXED

I took it one step further on the simplification scale.  I removed the NumberPicker and added the Decimals attribute to the DataGridNumericColumn element.  I feel like the format should work by default based on the data type of the field.  In any case I can now store a value of 2.101. 

```
<DataGridNumericColumn Field="@nameof(CaliberInfo.oal_max)" Decimals="3" Caption="OAL Max" DisplayFormat="{0:N6}" DisplayFormatProvider="@System.Globalization.CultureInfo.GetCultureInfo("en-EN")" Editable>
            ```
